### PR TITLE
[#167102335] Added selection in organization list component

### PR DESCRIPTION
**Pull request by laurapugliese**
Added selection in organization list component

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#167102335](https://www.pivotaltracker.com/story/show/167102335): Come CIT voglio poter selezionare gli enti visualizzati mediante la ricerca nella sezione "Locali"

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by laurapugliese** [comment]
This PR aims to insert the selection in the component that displays the list of organizations.

@Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)

selezione_compress

**Comment by Undermaken** [comment]
> This PR aims to insert the selection in the component that displays the list of organizations.
> 
> @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> 
> selezione_compress

Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):

> When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want

If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case

**Comment by laurapugliese** [comment]
> > This PR aims to insert the selection in the component that displays the list of organizations.
> > @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> > Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> > selezione_compress
> 
> @laurapugliese
> Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):
> 
> > When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want
> 
> If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case

Thanks @Undermaken!! It's ok if I leave the `ReadonlyArray<string>` type?

**Comment by Undermaken** [comment]
> > > This PR aims to insert the selection in the component that displays the list of organizations.
> > > @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> > > Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> > > selezione_compress
> > 
> > 
> > @laurapugliese
> > Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):
> > > When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want
> > 
> > 
> > If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case
> 
> Thanks @Undermaken!! It's ok if I leave the `ReadonlyArray<string>` type?

I didn't check the code. Anyway, why did you use a Set? Do you need special features provided by Set data structure?

**Comment by laurapugliese** [comment]
> > > > This PR aims to insert the selection in the component that displays the list of organizations.
> > > > @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> > > > Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> > > > selezione_compress
> > > 
> > > 
> > > @laurapugliese
> > > Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):
> > > > When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want
> > > 
> > > 
> > > If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case
> > 
> > 
> > Thanks @Undermaken!! It's ok if I leave the `ReadonlyArray<string>` type?
> 
> I didn't check the code. Anyway, why did you use a Set? Do you need special features provided by Set data structure?

Nothing special, but I used the HOC component for the selection used already in the messages where the Set is used and for convenience I wanted to keep the same type. But now I replaced it with an array. If you find something in the review that you don't like tell me quietly. ;)

**Comment by Undermaken** [comment]
> > > > > This PR aims to insert the selection in the component that displays the list of organizations.
> > > > > @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> > > > > Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> > > > > selezione_compress
> > > > 
> > > > 
> > > > @laurapugliese
> > > > Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):
> > > > > When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want
> > > > 
> > > > 
> > > > If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case
> > > 
> > > 
> > > Thanks @Undermaken!! It's ok if I leave the `ReadonlyArray<string>` type?
> > 
> > 
> > I didn't check the code. Anyway, why did you use a Set? Do you need special features provided by Set data structure?
> 
> Nothing special, but I used the HOC component for the selection used already in the messages where the Set is used and for convenience I wanted to keep the same type. But now I replaced it with an array. If you find something in the review that you don't like tell me quietly. ;)

**Comment by Undermaken** [comment]
> > > > > > This PR aims to insert the selection in the component that displays the list of organizations.
> > > > > > @Undermaken @fpersico in the reducer `organizationsFiscalCodesSelected` I initially used as `Option<Set<string>>` type (the same type used in the HOC component) but unfortunately the data was not stored when the app was closed. I replaced it with `ReadonlyArray<string>` and now it works.
> > > > > > Do you have any idea about? (In this commit 695c645I the version with the `Option<Set<string>>` type)
> > > > > > selezione_compress
> > > > > 
> > > > > 
> > > > > @laurapugliese
> > > > > Yes, redux-persist works as expected. This is from the [documentation](https://github.com/rt2zz/redux-persist):
> > > > > > When the state object gets persisted, it first gets serialized with JSON.stringify(). If parts of your state object are not mappable to JSON objects, the serialization process may transform these parts of your state in unexpected ways. For example, the javascript Set type does not exist in JSON. When you try to serialize a Set via JSON.stringify(), it gets converted to an empty object. Probably not what you want
> > > > > 
> > > > > 
> > > > > If you want to persist/rehydrate a **_Set_** you could check the **Transforms** section in the documentation that talks about this specific case
> > > > 
> > > > 
> > > > Thanks @Undermaken!! It's ok if I leave the `ReadonlyArray<string>` type?
> > > 
> > > 
> > > I didn't check the code. Anyway, why did you use a Set? Do you need special features provided by Set data structure?
> > 
> > 
> > Nothing special, but I used the HOC component for the selection used already in the messages where the Set is used and for convenience I wanted to keep the same type. But now I replaced it with an array. If you find something in the review that you don't like tell me quietly. ;)


Sorry for PR closing, bad button press 👎 
If you are using **Set** for representing _selectedItemIds_, I quite sure it is appropriate.
Set contains unique elements as ids are, indeed. And you are using  set's method: **has**, **add**, **delete** to make operations with it. If you don't use a **Set** you have to ensure the array will work as a **Set**

**Comment by gunzip** [comment]
not sure if its related but you may consider https://www.npmjs.com/package/json-set-map which provide a serializable set (a `Set ` that gets converted to an array when using `JSON.stringify`)

**Comment by Undermaken** [comment]
> not sure if its related but you may consider https://www.npmjs.com/package/json-set-map which provide a serializable set (a `Set ` that gets converted to an array when using `JSON.stringify`)

@gunzip Yes, it is. This will solve this case.
Anyway **redux-persist** provides _tranforms_ to control how data is stored and loaded back.
In this case, instead of a library, you could convert a _Set_ in an _array_ (which is a recognized data structure in JSON and it is an iterable) just before storing data, and then convert back the _array_ into a _Set_ when loading.

```javascript
import { createTransform } from 'redux-persist';

const SetTransform = createTransform(
  // transform state on its way to being serialized and persisted.
  (inboundState, key) => {
    // convert mySet to an Array.
    return { ...inboundState, mySet: [...inboundState.mySet] };
  },
  // transform state being rehydrated
  (outboundState, key) => {
    // convert mySet back to a Set.
    return { ...outboundState, mySet: new Set(outboundState.mySet) };
  },
  // define which reducers this transform gets called for.
  { whitelist: ['someReducer'] }
);

export default SetTransform;
```

We already did a [transformer](https://github.com/teamdigitale/io-app/blob/master/ts/store/transforms/dateISO8601Tranform.ts) for Date objects

File: ts/components/ChooserListSearch.tsx [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
   onSearchItemContainsText?: (item: T, searchText: string) => boolean;
   noSearchResultsSourceIcon?: ImageSourcePropType;
   noSearchResultsSubtitle?: string;
+  keyExtractor: (item: T) => string;
+  itemTitleExtractor: (item: T) => string;
+  selectedItemIds: Option<Set<string>>;
 };
 
-type Props<T> = OwnProps<T>;
+type ChooserListItemProps = "itemIconComponent" | "onPressItem";
```
Comment (Undermaken) [inline ts/components/ChooserListSearch.tsx]: what does it mean?
Could you add an explanation comment?

File: ts/components/ChooserListItem.tsx [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
+import variables from "../theme/variables";
 import customVariables from "../theme/variables";
+import IconFont from "./ui/IconFont";
 
 type Props = Readonly<{
-  title: string;
-  iconComponent?: React.ReactElement;
+  itemTitle: string;
+  itemId: string;
+  itemIconComponent?:
```
Comment (Undermaken) [inline ts/components/ChooserListItem.tsx]: A cleaner solution could be [Either](https://gcanti.github.io/fp-ts/modules/Either.ts.html)?

File: ts/components/ChooserListSearch.tsx (hunk 1 above)
Comment (laurapugliese) [reply ts/components/ChooserListSearch.tsx]: Are the properties that this component will have to pass to the ChooserListItem component and so I defined them the set of properties required and defined under `Pick<ComponentProps<typeof ChooserListItem>, ChooserListItemProps>`

File: ts/components/ChooserListSearch.tsx [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
 import customVariables from "../theme/variables";
 import ChooserList from "./ChooserList";
+import ChooserListItem from "./ChooserListItem";
 
 type OwnProps<T> = {
   listState: ReadonlyArray<T>;
   searchText: string;
-  keyExtractor: (item: T, index: number) => string;
-  renderItem: ListRenderItem<T>;
   onSearchItemContainsText?: (item: T, searchText: string) => boolean;
```
[the remark this answers is not in the document]
Comment (laurapugliese) [reply ts/components/ChooserListSearch.tsx]: Done in commit ffc0b71

File: ts/components/ChooserListContainer.tsx [hunk 4]
```diff
@@ -55,12 +67,8 @@ const DEBOUNCED_TIME = 300 as Millisecond;
```
[the remark this answers is not in the document]
Comment (laurapugliese) [reply ts/components/ChooserListContainer.tsx]: Done in commit f10ffcb

File: ts/components/ChooserListContainer.tsx [hunk 5]
```diff
@@ -256,3 +292,12 @@ export class ChooserListContainer<T> extends React.PureComponent<
     );
   }
 }
+
+type Without<T, K> = Pick<T, Exclude<keyof T, K>>;
```
[the remark this answers is not in the document]
Comment (francescopersico) [reply ts/components/ChooserListContainer.tsx]: @laurapugliese No need to create a new type there is already Omit

[the remark this answers is not in the document]
Comment (Undermaken) [reply ts/components/ChooserListContainer.tsx]: @francescopersico @laurapugliese What about this solution?

```javascript
export default <T extends {}>() => {
  const hocComponent = (props: Props<T>) => {
    const {
      selectedItemIds,
      toggleItemSelection,
      resetSelection,
      setSelectedItemIds,
      ...rest
    } = props;
    return (
      <ChooserListContainer<T>
        selectedItemIds={selectedItemIds}
        toggleItemSelection={toggleItemSelection}
        resetSelection={resetSelection}
        setSelectedItemIds={setSelectedItemIds}
        {...rest}
      />
    );
  };
  return withItemsSelection(hocComponent);
};
```

in place of 

```javascript
type Without<T, K> = Pick<T, Exclude<keyof T, K>>;
type ExternalProps<T> = Without<OtherProps<T>, "classes">;

type ChooserListContainerType = <T>(props: ExternalProps<T>) => any;

export default (withItemsSelection(
  ChooserListContainer
) as unknown) as ChooserListContainerType;
```

and in _ServicesHomeScreen.tsx_

```javascript
const OrganizationsList = createFactory<Organization>();
    this.props.showModal(
      <OrganizationsList
        items={allOrganizations}
        initialSelectedItemIds={some(new Set(organizationsSelected))}
        keyExtractor={(item: Organization) => item.fiscalCode}
        itemTitleExtractor={(item: Organization) => item.name}
        itemIconComponent={(fiscalCode: string) =>
          this.renderOrganizationLogo(fiscalCode)
        }
        onCancel={hideModal}
        onSave={this.onSaveAreasOfInterest}
        isRefreshEnabled={false}
        isSearchEnabled={true}
        onSearchItemContainsText={this.organizationContainsText}
        noSearchResultsSourceIcon={require("../../../img/services/icon-no-places.png")}
        noSearchResultsSubtitle={I18n.t("services.areasOfInterest.searchEmpty")}
      />
    );
```

[the remark this answers is not in the document]
Comment (laurapugliese) [reply ts/components/ChooserListContainer.tsx]: > Omit

Initially I had used Omit (695c645) but failed the tests

[the remark this answers is not in the document]
Comment (laurapugliese) [reply ts/components/ChooserListContainer.tsx]: @fpersico Initially I used Omit (695c645) but failed the tests

File: ts/components/ChooserListItem.tsx (hunk 2 above)
Comment (laurapugliese) [reply ts/components/ChooserListItem.tsx]: Done in commit 1a7b88d

File: ts/components/ChooserListContainer.tsx [hunk 6]
```diff
@@ -5,35 +5,47 @@ import { Millisecond } from "italia-ts-commons/lib/units";
...
-  renderItem: ListRenderItem<T>;
+  initialSelectedItemIds: Option<Set<string>>;
+  keyExtractor: (item: T) => string;
+  itemTitleExtractor: (item: T) => string;
   onCancel: () => void;
+  onSave: (selectedItemIds: Option<Set<string>>) => void;
   isRefreshEnabled: boolean;
   isRefreshing?: boolean;
   onRefresh?: () => void;
   isSearchEnabled: boolean;
```
Comment (Undermaken) [inline ts/components/ChooserListContainer.tsx]: What do you think about removing **isSearchEnabled** prop?
What does it happen if you set **isSearchEnabled : true** and **matchingTextPredicate : undefined**? Does it make sense?
You could say a search is active when you have a predicate set.

File: ts/components/services/ServicesLocalHeader.tsx [hunk 7]
```diff
@@ -38,10 +42,33 @@ const styles = StyleSheet.create({
...
+export class ServicesLocalHeader extends React.PureComponent<Props> {
   public render() {
-    const { paddingForAnimation, onAddAreasOfInterestPress } = this.props;
-    return (
+    const {
+      paddingForAnimation,
+      onChooserAreasOfInterestPress,
+      organizationsFiscalCodesSelected
+    } = this.props;
+    const isOrganizationsFiscalCodesSelected =
```
Comment (Undermaken) [inline ts/components/services/ServicesLocalHeader.tsx]:
```suggestion
    const isOrganizationsFiscalCodesSelected = organizationsFiscalCodesSelected.fold(
      false,
      _ => _.size > 0
    );
```

Comment (laurapugliese) [reply ts/components/services/ServicesLocalHeader.tsx]: Done in commit 0c15e67

File: ts/components/ChooserListContainer.tsx (hunk 6 above)
Comment (laurapugliese) [reply ts/components/ChooserListContainer.tsx]: Yes, I used matchingTextPredicate to enable the search. Done in commit 0c15e67

**Comment by Undermaken** [comment]
This PR will be merged. [This point](https://github.com/teamdigitale/io-app/pull/1240#discussion_r309321651) will be recovered in another story

**Pull request merged**